### DISCUSSION TITLE
Lots of idle sessions with sqlalchemy+psycopg

### DISCUSSION THREAD

**Discussion by aranvir** [sqlalchemy/sqlalchemy#11643] [Usage Questions] [answered]
I did some load testing for my litestar application (2 uvicorn workers) and realized that I would hit the postgres connection limit relatively easily. This is set to 100 by default, but what's confusing to me is that the engine default settings should not even allow 50 connections in the first place.

I tested multiple configurations, but the one below I ended on. For load testing I use locust with 10 users and 0.5 user/s ramp up. The test runs for 1 minute. The test is simple: log in and then just query 1 endpoint. What I would expect is to see a sudden ramp up to 10-20 sessions, then between 10-20 sessions during the test and at the end returning to 10 idle sessions (2x pool size). I appended a snippet from pg admin below - instead the idle sessions ramp up to 61. And stay there after the test. 

Before I simplified the locust test, the session behavior was a bit more "oscillating", e.g., 22, 54, 35, 54, 44, 80, 42, 96, 45, 55, 48. (see second screenshot)
So, something is cleaning up those sessions eventually - just not how I thought it would.

Unfortunately I cannot recreate the problem with a simple example. 

Some version information:
Python 3.11
Litestar 2.9.0
psycopg 3.1.18
uvicorn 0.30.3
Postgres 16

```python
    engine_config = EngineConfig(
        pool_pre_ping=True,
        pool_size=5,
        # pool_recycle=3600,
        max_overflow=5,
        pool_use_lifo=True,
        connect_args={
            "application_name": f"LITESTAR-{os.getpid()}"
        }
    )
    session_config = AsyncSessionConfig(expire_on_commit=False)
    sqlalchemy_config = SQLAlchemyAsyncConfig(
        connection_string=conn or get_connection_string(),
        session_config=session_config,
        engine_config=engine_config
    )  # Create 'async_session' dependency.
    # This will take care of disposing the async engine at the end of the app lifespan
    sqlalchemy_init_plugin = SQLAlchemyInitPlugin(config=sqlalchemy_config)
```

![image](https://github.com/user-attachments/assets/c4c21969-b383-44c2-a842-ec21b3a68ff2)

![image](https://github.com/user-attachments/assets/2857bdc6-83a4-45b3-9ae4-3774dc74727b)

**Comment by aranvir** [comment, marked as answer]
it was a user issue - in my authentication layer I misunderstood a library function which caused an engine to be created per sessionmaker call instead of calling the existing sessionmaker and engine.

**Reply by CaselIT** [reply]
Ok!
